onnxruntime/core/optimizer/reshape_fusion.cc

```diff
@@ -77,84 +80,57 @@ bool ReshapeFusion::Fuse_Subgraph1(Node& reshape, Graph& graph, const logging::L
   }
 
   auto concat_input_count = concat.InputArgCount().front();
-  if (concat_input_count < 3 || concat_input_count > 4 || concat.GetOutputEdgesCount() > 1) {
+  if (concat.GetOutputEdgesCount() > 1) {
     return false;
   }
 
-  // path 1: [Root] --> Shape --> Gather(indices=0) --> Unsqueeze (axes=0) --> Concat [input 0]
-  std::vector<graph_utils::EdgeEndToMatch> parent_path{
-      {0, 0, "Unsqueeze", {1, 11}, kOnnxDomain},
-      {0, 0, "Gather", {1, 11}, kOnnxDomain},
-      {0, 0, "Shape", {1}, kOnnxDomain}};
-
-  std::vector<const Node::EdgeEnd*> edges;
-  if (!graph_utils::FindPath(concat, true, parent_path, edges, logger)) {
-    return false;
-  }
-
-  const Node& unsqueeze_1 = edges[0]->GetNode();
-  const Node& gather_1 = edges[1]->GetNode();
-  const Node& shape_1 = edges[2]->GetNode();
-  if (unsqueeze_1.GetOutputEdgesCount() != 1 || gather_1.GetOutputEdgesCount() != 1 || shape_1.GetOutputEdgesCount() != 1) {
-    return false;
-  }
-
-  if (graph_utils::GetInputNode(shape_1, 0) != p_root) {
-    return false;
-  }
-
-  std::vector<int64_t> axes;
-  if (!(graph_utils::GetRepeatedNodeAttributeValues(unsqueeze_1, "axes", axes) && axes.size() == 1 && axes[0] == 0)) {
-    return false;
-  }
-
-  if (!optimizer_utils::IsInitializerWithExpectedValue(graph, *(gather_1.InputDefs()[1]), int64_t(0), false)) {
-    return false;
-  }
-
-  // path 2: [Root] --> Shape --> Gather(indices=1) --> Unsqueeze (axes=0) --> Concat [input 1]
-  std::vector<graph_utils::EdgeEndToMatch> parent_path2 {
-      {0, 1, "Unsqueeze", {1, 11}, kOnnxDomain},
-      {0, 0, "Gather", {1, 11}, kOnnxDomain},
-      {0, 0, "Shape", {1}, kOnnxDomain}};
-
-  if (!graph_utils::FindPath(concat, true, parent_path2, edges, logger)) {
-    return false;
-  }
-
-  const Node& unsqueeze_2 = edges[0]->GetNode();
-  const Node& gather_2 = edges[1]->GetNode();
-  const Node& shape_2 = edges[2]->GetNode();
-  if (unsqueeze_2.GetOutputEdgesCount() != 1 || gather_2.GetOutputEdgesCount() != 1 || shape_2.GetOutputEdgesCount() != 1) {
-    return false;
-  }
+  std::vector<int64_t> shape_value;
+  shape_value.reserve(concat_input_count);
+  // Used to keep the following nodes in the order of their potential removal.
+  enum class NodeType { Unsqueeze, Gather, Shape };
+  std::set<std::pair<NodeType, NodeIndex>> candidates_for_removal;
```

Comment:
candidates_for_removal can be std::vector<NodeIndex>. Then append node index of unsqueeze, gather, shape nodes in the order.

Reply:
This will not work correctly in case of sharing, e.g. when different Gather nodes refer to the same Shape node.

Reply:
That's right. That need check node index does not exist in the vector before appending (to avoid duplicated node index).

Reply:
Then the vector will be in the order [unsqueeze1, gather1, shared_shape, unsqueeze2, gather2], which is incorrect because we want to remove nodes in reverse topological order (unsqueeze2 and gather2 must come before shared_shape).